azure-devops/run-ci-builds.ps1

```diff
@@ -3,31 +3,80 @@ param (
-    [string] $SourceVersion
+    [Parameter(Mandatory)] [string] $SourceVersion,
+    [Parameter(Mandatory)] [string] $ManifestLink,
+    [Parameter(Mandatory)] [bool] $WaitForBuilds,
+    [string] $ToolVersions,
+    [UInt32] $RetryIntervalSec = 30,
+    [UInt32] $RetryCount = 20
 )
 
 Import-Module (Join-Path $PSScriptRoot "azure-devops-api.ps1")
 Import-Module (Join-Path $PSScriptRoot "build-info.ps1")
 
+function Get-ToolVersions {
+    param (
+        [Parameter(Mandatory)] [string] $ManifestLink,
+        [Parameter(Mandatory)] [UInt32] $RetryIntervalSec,
+        [Parameter(Mandatory)] [UInt32] $Retries,
+        [string] $ToolVersions
+    )
+    
+    [string[]] $versionsList = @()
+    if ($ToolVersions) {
+        $versionsList = $ToolVersions.Split(',')
+    } else {
+        Write-Host "Get the list of releases from $ManifestLink"
+        $releases = Invoke-RestMethod $ManifestLink -MaximumRetryCount $Retries -RetryIntervalSec $RetryIntervalSec
+        $versionsList = $releases.version
+    }
+
+    return $versionsList
+}
+
 function Queue-Builds {
     param (
         [Parameter(Mandatory)] [AzureDevOpsApi] $AzureDevOpsApi,
-        [Parameter(Mandatory)] [string] $ToolVersions,
+        [Parameter(Mandatory)] [string[]] $ToolVersions,
         [Parameter(Mandatory)] [string] $SourceBranch,
         [Parameter(Mandatory)] [string] $SourceVersion,
-        [Parameter(Mandatory)] [string] $DefinitionId
+        [Parameter(Mandatory)] [UInt32] $DefinitionId,
+        [Parameter(Mandatory)] [UInt32] $RetryIntervalSec,
+        [Parameter(Mandatory)] [UInt32] $Retries
     )
 
     [BuildInfo[]]$queuedBuilds = @()
 
-    $ToolVersions.Split(',') | ForEach-Object { 
+    $ToolVersions | ForEach-Object { 
         $version = $_.Trim()
-        Write-Host "Queue build for $version..."
-        $queuedBuild = $AzureDevOpsApi.QueueBuild($version, $SourceBranch, $SourceVersion, $DefinitionId)
-        $buildInfo = Get-BuildInfo -AzureDevOpsApi $AzureDevOpsApi -Build $queuedBuild
-        Write-Host "Queued build: $($buildInfo.Link)"
+        
+        while ($Retries -gt 0)
+        {
+            try
+            {
+                Write-Host "Queue build for $version..."
+                $queuedBuild = $AzureDevOpsApi.QueueBuild($version, $SourceBranch, $SourceVersion, $DefinitionId)
+                $buildInfo = Get-BuildInfo -AzureDevOpsApi $AzureDevOpsApi -Build $queuedBuild
+                Write-Host "Queued build: $($buildInfo.Link)"
+                break
+            }
+            catch
+            {
+                Write-Host "There is an error during build starting:`n $_"
+                $Retries--
+            
+                if ($Retries -eq 0)
+                {
+                    Write-Host "Build can't be queued, please try later"
+                    exit 1
+                }
+            
+                Write-Host "Waiting 30 seconds before retrying. Retries left: $Retries"
```

Comment:
It is too difficult for read the code.
Could you please move retry logic to helper function like `Invoke-RestMethodWithRetry`

Reply:
Can we also add retry logic to Wait-Builds method?

Reply:
Good point, thanks! I've moved retry logic to the `AzureDevOpsApi` class